--- avidemux.SlackBuild.orig	2015-03-07 18:39:25.000000000 -0500
+++ avidemux.SlackBuild	2015-03-08 23:05:44.695226763 -0400
@@ -26,7 +26,7 @@
 
 PRGNAM=avidemux
 VERSION=${VERSION:-2.6.8}
-BUILD=${BUILD:-1}
+BUILD=${BUILD:-2}
 TAG=${TAG:-_SBo}
 
 DOCS="AUTHORS License.txt COPYING scripts/automkv.py"
@@ -73,53 +73,63 @@
  \( -perm 666 -o -perm 664 -o -perm 640 -o -perm 600 -o -perm 444 \
   -o -perm 440 -o -perm 400 \) -exec chmod 644 {} \;
 
+WITH_CLI=${WITH_CLI:-YES}
+WITH_QT4=${WITH_QT4:-YES}
+WITH_GTK=${WITH_GTK:-NO}
+
 mkdir -p build_core
 cd build_core
-  cmake \
-    -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
-    -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
-    -DCMAKE_INSTALL_PREFIX=/usr \
-    -DFAKEROOT="$PKG" \
-    ../avidemux_core
-  make
-  make install DESTDIR=$PKG
+cmake \
+  -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
+  -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
+  -DCMAKE_INSTALL_PREFIX=/usr \
+  -DFAKEROOT="$PKG" \
+  ../avidemux_core
+make -j 1
+make -j 1 install DESTDIR=$PKG
 cd ..
 
-mkdir -p build_qt4
-cd build_qt4
+if [ "x$WITH_QT4" = "xYES" ]; then
+  mkdir -p build_qt4
+  cd build_qt4
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_INSTALL_PREFIX=/usr \
     -DFAKEROOT="$PKG" \
     ../avidemux/qt4
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
-mkdir -p build_gtk
-cd build_gtk
+if [ "x$WITH_GTK" = "xYES" ]; then
+  mkdir -p build_gtk
+  cd build_gtk
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_INSTALL_PREFIX=/usr \
     -DFAKEROOT="$PKG" \
     ../avidemux/gtk
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
-mkdir -p build_cli
-cd build_cli
+if [ "x$WITH_CLI" = "xYES" ]; then
+  mkdir -p build_cli
+  cd build_cli
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_INSTALL_PREFIX=/usr \
     -DFAKEROOT="$PKG" \
     ../avidemux/cli
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
 mkdir -p build_plugins_common
 cd build_plugins_common
@@ -131,12 +142,13 @@
     -DPLUGIN_UI=COMMON \
     -DFAKEROOT="$PKG" \
     ../avidemux_plugins
-  make
-  make install DESTDIR=$PKG
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
 cd ..
 
-mkdir -p build_plugins_qt4
-cd build_plugins_qt4
+if [ "x$WITH_QT4" = "xYES" ]; then
+  mkdir -p build_plugins_qt4
+  cd build_plugins_qt4
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
@@ -145,12 +157,14 @@
     -DPLUGIN_UI=QT4 \
     -DFAKEROOT="$PKG" \
     ../avidemux_plugins
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
-mkdir -p build_plugins_gtk
-cd build_plugins_gtk
+if [ "x$WITH_GTK" = "xYES" ]; then
+  mkdir -p build_plugins_gtk
+  cd build_plugins_gtk
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
@@ -159,12 +173,14 @@
     -DPLUGIN_UI=GTK \
     -DFAKEROOT="$PKG" \
     ../avidemux_plugins
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
-mkdir -p build_plugins_cli
-cd build_plugins_cli
+if [ "x$WITH_CLI" = "xYES" ]; then
+  mkdir -p build_plugins_cli
+  cd build_plugins_cli
   cmake \
     -DCMAKE_C_FLAGS:STRING="$SLKCFLAGS" \
     -DCMAKE_CXX_FLAGS:STRING="$SLKCFLAGS" \
@@ -173,15 +189,21 @@
     -DPLUGIN_UI=CLI \
     -DFAKEROOT="$PKG" \
     ../avidemux_plugins
-  make
-  make install DESTDIR=$PKG
-cd ..
+  make -j 1
+  make -j 1 install DESTDIR=$PKG
+  cd ..
+fi
 
 mkdir -p $PKG/usr/share/applications
 mkdir -p $PKG/usr/share/pixmaps
-install -m644 avidemux2.desktop $PKG/usr/share/applications/avidemux2_gtk.desktop
-cat avidemux2.desktop | sed -e 's/_gtk/_qt4/' \
-  > $PKG/usr/share/applications/avidemux2_qt4.desktop
+if [ "x$WITH_QT4" = "xYES" ]; then
+	cat avidemux2.desktop | sed -e 's/avidemux2$/avidemux/' -e 's/2_gtk/3_qt/' >\
+		$PKG/usr/share/applications/avidemux_qt.desktop
+fi
+if [ "x$WITH_GTK" = "xYES" ]; then
+	cat avidemux2.desktop | sed -e 's/avidemux2$/avidemux/' -e 's/2_gtk/3_gtk/' >\
+		$PKG/usr/share/applications/avidemux_gtk.desktop
+fi
 install -m644 avidemux_icon.png $PKG/usr/share/pixmaps/avidemux.png
 for sz in 16x16 32x32 48x48 64x64 ; do
   mkdir -p $PKG/usr/share/icons/hicolor/$sz/apps
@@ -190,14 +212,11 @@
 done
 
 mkdir -p $PKG/usr/man/man1
-install -m644 man/avidemux.1 $PKG/usr/man/man1/avidemux.1
+gzip -c man/avidemux.1 > $PKG/usr/man/man1/avidemux.1.gz
 
 find $PKG -print0 | xargs -0 file | grep -e "executable" -e "shared object" | grep ELF \
   | cut -f 1 -d : | xargs strip --strip-unneeded 2> /dev/null || true
 
-find $PKG/usr/man -type f -exec gzip -9 {} \;
-for i in $( find $PKG/usr/man -type l ) ; do ln -s $( readlink $i ).gz $i.gz ; rm $i ; done
-
 mkdir -p $PKG/usr/doc/${PRGNAM}-${VERSION}
 cp -a $DOCS $PKG/usr/doc/${PRGNAM}-${VERSION}
 cat $CWD/$PRGNAM.SlackBuild > $PKG/usr/doc/${PRGNAM}-${VERSION}/$PRGNAM.SlackBuild
