diff -Naur a/configure.ac b/configure.ac
--- a/configure.ac	2014-01-05 15:21:03.000000000 -0500
+++ b/configure.ac	2014-01-26 23:37:32.558397620 -0500
@@ -51,7 +51,7 @@
                   ],
                   [AC_CHECK_LIB([crypto],
                       [EVP_CipherInit_ex],
-                      [AC_CHECK_HEADERS([openssl/bn.h openssl/aes.h openssl/sha.h],
+                      [AC_CHECK_HEADERS([openssl/bn.h openssl/aes.h openssl/sha.h pthread.h],
                                         [], AC_MSG_ERROR([OpenSSL headers not found.]))
                        LIBS="-lcrypto $LIBS"
                        openssl="true"
diff -Naur a/src/Makefile.am b/src/Makefile.am
--- a/src/Makefile.am	2013-06-27 08:04:12.000000000 -0400
+++ b/src/Makefile.am	2014-01-26 23:36:08.857397620 -0500
@@ -22,7 +22,7 @@
 
 if SRTP_OPENSSL
 srtp_src_o =    ccrtp/crypto/openssl/hmac.cpp \
-        ccrtp/crypto/openssl/SrtpSymCrypto.cpp
+        ccrtp/crypto/openssl/SrtpSymCrypto.cpp ccrtp/crypto/twofish.c
 endif
 
 if SRTP_GCRYPT
